### PR TITLE
Async coll perm graph revision

### PR DESCRIPTION
**Pull request by escherize**
Pursuant to https://github.com/metabase/metabase/issues/39997

Improves collection graph update performance, especially on problematic instances that have 1000+ collections and perm groups, by making the costly process of writing the potentially gigantic revision json blob to the db into an async process.

An alternative solution is to remove the collection perm graph revision capturing entirely. We still might do that since metabase/metabase does not use the perm graph revision for anything. But it may be nice to have for users and/or support. If it's needed, we could consider a diff-based approach instead.

### CODE REVIEW

File: src/metabase/models/collection/graph.clj [hunk 1]
```diff
@@ -170,13 +170,15 @@
          new-perms          (into {} (for [[group-id collection-id->perms] new-perms]
                                       [group-id (select-keys collection-id->perms (keys (get old-perms group-id)))]))
          [diff-old changes] (data/diff old-perms new-perms)]
-     (data-perms.graph/log-permissions-changes diff-old changes)
```
Comment (escherize) [inline src/metabase/models/collection/graph.clj:173]: We were logging before checking that the revision numbers were correct. That seems wrong, so I change the order that this happens too.

**Review by johnswanson — CHANGES_REQUESTED** [review]
This makes me a bit nervous - it doesn't provide the same concurrency guarantees as before.

Say `save-perms-revision!` takes 60s to complete. We'll update the permissions graph, but the revision won't be updated for a minute. So e.g. say I am on revision 47 and make a change. The endpoint returns quickly. At this moment, someone else makes a different change to permissions - also with revision 47. `check-revision-numbers` won't throw because as far as it knows we're still on revision 47. It'll successfully change the permissions graph - then sometime in the future we'll get an exception when we try to insert two revisions with the same ID (one will be lost).

In the end we'll have a single new revision (48) and I think it'll have the wrong diff.

One option would be to decouple the revision number from the actual revision write itself (maybe we could insert it and then fill it in later?) but I'm not really sure.

**Comment by escherize** [comment]
Good call, I will decouple incrementing the revision number (fast, and should be sync) from writing the perm-graph (slow and will be async).

File: src/metabase/models/collection/graph.clj [hunk 2]
```diff
@@ -153,6 +152,22 @@
...
+(defn create-perms-revision!
+  "Increments the current revision number and writes it to the database. This lets us track the permissions graph
+  revision number, which is used for consistency checks when updating the graph."
+  [current-revision-number]
+  (when api/*current-user-id*
+    (first (t2/insert-returning-instances! :model/CollectionPermissionGraphRevision
+                                           :id      (inc current-revision-number)
+                                           :user_id api/*current-user-id*
+                                           :before ""
+                                           :after ""))))
```
Comment (escherize) [inline src/metabase/models/collection/graph.clj:164]: I almost wrote a migration here, but was swayed by: https://github.com/metabase/metabase/blob/async-coll-perm-graph-revision/resources/migrations/001_update_migrations.yaml#L8267

**Review by johnswanson — APPROVED** [review]
Nice, LGTM!

File: src/metabase/models/collection/graph.clj [hunk 3]
```diff
@@ -153,6 +152,23 @@
   version."
   metabase-enterprise.audit-app.permissions [_ _] ::noop)
 
+(defn create-perms-revision!
+  "Increments the current revision number and writes it to the database. This lets us track the permissions graph
+  revision number, which is used for consistency checks when updating the graph."
+  [current-revision-number]
+  (when api/*current-user-id*
+    (first (t2/insert-returning-instances! :model/CollectionPermissionGraphRevision
+                                           :id      (inc current-revision-number)
```
Comment (noahmoss) [inline src/metabase/models/collection/graph.clj:161]: nit
```suggestion
                                           :id (inc current-revision-number)
```

**Review by noahmoss — COMMENTED** [review]
Filling in the before/after async makes sense to me. For testing, I wonder if we can just deref the future returned from `update-graph!` so that the test blocks until the data is written? Is there anything that requires the wait-until approach? Hard-coded sleeps in tests just make me nervous.

**Pull request merged**